File: baekjoon_online_judge/34753/solution.py
import sys
input = sys.stdin.readline
import heapq

N = int(input())

A = [list(map(int, input().split())) for _ in range(N)]

# 줄 ID 부여
# 0 ~ N-1: 행
# N ~ 2N-1: 열
# 2N: diag ↘
# 2N+1: anti-diag ↙
ROW = 0
COL = N
DIA = 2 * N
ADI = 2 * N + 1

total_lines = 2 * N + 2

# 각 줄별 남은 비용(초기)
remain = [0] * total_lines

# 행
for i in range(N):
    remain[ROW + i] = sum(A[i])
# 열
for j in range(N):
    s = 0
    for i in range(N):
        s += A[i][j]
    remain[COL + j] = s
# diag ↘ (i == j)
s = 0
for i in range(N):
    s += A[i][i]
remain[DIA] = s
# anti-diag ↙ (i + j == N-1)
s = 0
for i in range(N):
    s += A[i][N - 1 - i]
remain[ADI] = s

# 각 줄의 타입/번호 (tie-breaking용)
# type order: row -> col -> diag -> anti
def line_order(lineid):
    if lineid < COL:
        return (0, lineid)   # row
    elif lineid < DIA:
        return (1, lineid - COL)  # col
    elif lineid == DIA:
        return (2, 0)        # diag
    else:
        return (3, 0)        # anti

# min-heap: (remain_cost, type, idx, lineid)
# type, idx는 tie-breaking을 위해 저장
heap = []
for lid in range(total_lines):
    typ, num = line_order(lid)
    heapq.heappush(heap, (remain[lid], typ, num, lid))

# 줄이 끝났는지
done = [False] * total_lines

# 결과 저장: k빙고 달성 시간
answer = [0] * (total_lines + 1)

cur_bingo = 0
time = 0

# 각 칸이 속한 줄 리스트를 미리 만들어두기
# 최대 4개의 줄(row, col, diag, anti)
belongs = [[None] * N for _ in range(N)]  # 칸 당 (row, col, diag?, anti?)
for i in range(N):
    for j in range(N):
        lst = []
        lst.append(ROW + i)
        lst.append(COL + j)
        if i == j:
            lst.append(DIA)
        if i + j == N - 1:
            lst.append(ADI)
        belongs[i][j] = lst

# 메인 루프: 2N+2 줄이 모두 끝날 때까지
finished_count = 0
while finished_count < total_lines:
    rc, typ, num, lid = heapq.heappop(heap)
    
    # lazy check
    if done[lid]:
        continue
    if rc != remain[lid]:
        continue

    # 이 줄을 끝냄
    done[lid] = True
    finished_count += 1

    time += rc
    cur_bingo += 1
    answer[cur_bingo] = time

    # lid 줄의 모든 칸을 순회하며 업데이트
    if lid < COL:
        # row lid: row = lid
        i = lid
        for j in range(N):
            if A[i][j] != 0:
                val = A[i][j]
                # 이 칸은 이제 사용됨 → 0으로 바꿈
                A[i][j] = 0
                # 이 칸이 속한 다른 줄들 업데이트
                for other in belongs[i][j]:
                    if not done[other]:
                        remain[other] -= val
                        otyp, onum = line_order(other)
                        heapq.heappush(heap, (remain[other], otyp, onum, other))

    elif lid < DIA:
        # col lid-N
        j = lid - COL
        for i in range(N):
            if A[i][j] != 0:
                val = A[i][j]
                A[i][j] = 0
                for other in belongs[i][j]:
                    if not done[other]:
                        remain[other] -= val
                        otyp, onum = line_order(other)
                        heapq.heappush(heap, (remain[other], otyp, onum, other))

    elif lid == DIA:
        # diag ↘, i==j
        for i in range(N):
            j = i
            if A[i][j] != 0:
                val = A[i][j]
                A[i][j] = 0
                for other in belongs[i][j]:
                    if not done[other]:
                        remain[other] -= val
                        otyp, onum = line_order(other)
                        heapq.heappush(heap, (remain[other], otyp, onum, other))

    else:
        # anti ↙, i+j==N-1
        for i in range(N):
            j = N - 1 - i
            if A[i][j] != 0:
                val = A[i][j]
                A[i][j] = 0
                for other in belongs[i][j]:
                    if not done[other]:
                        remain[other] -= val
                        otyp, onum = line_order(other)
                        heapq.heappush(heap, (remain[other], otyp, onum, other))

# 출력
out = []
for k in range(1, total_lines + 1):
    out.append(str(answer[k]))

print("\n".join(out))

File: baekjoon_online_judge/34753/solution.cpp
#include <bits/stdc++.h>
using namespace std;

static long long A[1501][1501];
static long long remain_cost[3005];
static bool done[3005];

int N;

struct Node {
    long long cost;
    int type;
    int num;
    int id;
    bool operator<(Node const& other) const {
        if (cost != other.cost) return cost > other.cost;  // min-heap
        if (type != other.type) return type > other.type;
        return num > other.num;
    }
};

pair<int,int> lineOrder(int lid, int N){
    if (lid < N) return {0, lid};           // row
    else if (lid < 2*N) return {1, lid-N};  // col
    else if (lid == 2*N) return {2, 0};     // diag
    else return {3, 0};                     // anti
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> N;
    for(int i=0;i<N;i++){
        for(int j=0;j<N;j++){
            cin >> A[i][j];
        }
    }

    int ROW = 0;
    int COL = N;
    int DIA = 2*N;
    int ADI = 2*N+1;
    int total_lines = 2*N + 2;

    // 초기 남은 비용 계산
    for(int i=0;i<N;i++){
        long long s=0;
        for(int j=0;j<N;j++) s += A[i][j];
        remain_cost[ROW+i] = s;
    }
    for(int j=0;j<N;j++){
        long long s=0;
        for(int i=0;i<N;i++) s += A[i][j];
        remain_cost[COL+j] = s;
    }
    long long s=0;
    for(int i=0;i<N;i++) s += A[i][i];
    remain_cost[DIA] = s;
    s=0;
    for(int i=0;i<N;i++) s += A[i][N-1-i];
    remain_cost[ADI] = s;

    priority_queue<Node> pq;
    for(int lid=0; lid<total_lines; lid++){
        auto o = lineOrder(lid, N);
        pq.push({remain_cost[lid], o.first, o.second, lid});
    }

    vector<long long> ans(total_lines+1);
    long long cur_time = 0;
    int bingo = 0;

    while (bingo < total_lines){
        Node nd;
        do {
            nd = pq.top(); pq.pop();
        } while(done[nd.id] || nd.cost != remain_cost[nd.id]);

        int lid = nd.id;
        done[lid] = true;

        cur_time += remain_cost[lid];
        bingo++;
        ans[bingo] = cur_time;

        // 해당 줄 완성 → 포함된 모든 칸의 값만큼 다른 줄 감소
        if (lid < COL){ // row
            int i = lid;
            for(int j=0;j<N;j++){
                long long val = A[i][j];
                if(val == 0) continue;
                A[i][j] = 0;
                // row
                if(!done[ROW+i]){
                    remain_cost[ROW+i] -= val;
                    auto o = lineOrder(ROW+i,N);
                    pq.push({remain_cost[ROW+i], o.first, o.second, ROW+i});
                }
                // col
                if(!done[COL+j]){
                    remain_cost[COL+j] -= val;
                    auto o = lineOrder(COL+j,N);
                    pq.push({remain_cost[COL+j], o.first, o.second, COL+j});
                }
                // diag
                if(i == j){
                    if(!done[DIA]){
                        remain_cost[DIA] -= val;
                        auto o = lineOrder(DIA,N);
                        pq.push({remain_cost[DIA], o.first, o.second, DIA});
                    }
                }
                // anti
                if(i + j == N-1){
                    if(!done[ADI]){
                        remain_cost[ADI] -= val;
                        auto o = lineOrder(ADI,N);
                        pq.push({remain_cost[ADI], o.first, o.second, ADI});
                    }
                }
            }
        }
        else if (lid < DIA){ // col
            int j = lid - COL;
            for(int i=0;i<N;i++){
                long long val = A[i][j];
                if(val == 0) continue;
                A[i][j] = 0;
                if(!done[ROW+i]){
                    remain_cost[ROW+i] -= val;
                    auto o = lineOrder(ROW+i,N);
                    pq.push({remain_cost[ROW+i], o.first, o.second, ROW+i});
                }
                if(!done[COL+j]){
                    remain_cost[COL+j] -= val;
                    auto o = lineOrder(COL+j,N);
                    pq.push({remain_cost[COL+j], o.first, o.second, COL+j});
                }
                if(i == j){
                    if(!done[DIA]){
                        remain_cost[DIA] -= val;
                        auto o = lineOrder(DIA,N);
                        pq.push({remain_cost[DIA], o.first, o.second, DIA});
                    }
                }
                if(i + j == N-1){
                    if(!done[ADI]){
                        remain_cost[ADI] -= val;
                        auto o = lineOrder(ADI,N);
                        pq.push({remain_cost[ADI], o.first, o.second, ADI});
                    }
                }
            }
        }
        else if (lid == DIA){
            for(int i=0;i<N;i++){
                int j=i;
                long long val = A[i][j];
                if(val == 0) continue;
                A[i][j] = 0;
                if(!done[ROW+i]){
                    remain_cost[ROW+i] -= val;
                    auto o = lineOrder(ROW+i,N);
                    pq.push({remain_cost[ROW+i], o.first, o.second, ROW+i});
                }
                if(!done[COL+j]){
                    remain_cost[COL+j] -= val;
                    auto o = lineOrder(COL+j,N);
                    pq.push({remain_cost[COL+j], o.first, o.second, COL+j});
                }
                if(!done[DIA]){
                    remain_cost[DIA] -= val;
                    auto o = lineOrder(DIA,N);
                    pq.push({remain_cost[DIA], o.first, o.second, DIA});
                }
                if(i + j == N-1){
                    if(!done[ADI]){
                        remain_cost[ADI] -= val;
                        auto o = lineOrder(ADI,N);
                        pq.push({remain_cost[ADI], o.first, o.second, ADI});
                    }
                }
            }
        }
        else { // ADI
            for(int i=0;i<N;i++){
                int j = N-1-i;
                long long val = A[i][j];
                if(val == 0) continue;
                A[i][j] = 0;
                if(!done[ROW+i]){
                    remain_cost[ROW+i] -= val;
                    auto o = lineOrder(ROW+i,N);
                    pq.push({remain_cost[ROW+i], o.first, o.second, ROW+i});
                }
                if(!done[COL+j]){
                    remain_cost[COL+j] -= val;
                    auto o = lineOrder(COL+j,N);
                    pq.push({remain_cost[COL+j], o.first, o.second, COL+j});
                }
                if(i == j){
                    if(!done[DIA]){
                        remain_cost[DIA] -= val;
                        auto o = lineOrder(DIA,N);
                        pq.push({remain_cost[DIA], o.first, o.second, DIA});
                    }
                }
                if(!done[ADI]){
                    remain_cost[ADI] -= val;
                    auto o = lineOrder(ADI,N);
                    pq.push({remain_cost[ADI], o.first, o.second, ADI});
                }
            }
        }
    }

    for(int i=1;i<=total_lines;i++){
        cout << ans[i] << "\n";
    }
}